compact_str/src/repr/mod.rs

```diff
@@ -13,29 +13,65 @@ mod boxed;
 mod discriminant;
 mod heap;
 mod inline;
+mod nonmax;
 
 use discriminant::{
     Discriminant,
     DiscriminantMask,
 };
 use heap::HeapString;
 use inline::InlineString;
+use nonmax::NonMaxU8;
 
 const MAX_SIZE: usize = std::mem::size_of::<String>();
-const EMPTY: Repr = Repr {
-    inline: InlineString::new_const(""),
-};
+const PADDING_SIZE: usize = MAX_SIZE - std::mem::size_of::<u8>();
+const EMPTY: Repr = Repr::from_inline(InlineString::new_const(""));
 
 // Used as a discriminant to identify different variants
-pub const HEAP_MASK: u8 = 0b11111111;
-
-pub union Repr {
+pub const HEAP_MASK: u8 = 0b11111110;
+
+#[repr(C)]
+pub struct Repr(
+    *const u8,
+    usize,
+    #[cfg(target_pointer_width = "64")] u32,
+    u16,
+    u8,
+    NonMaxU8,
+);
```

Comment:
How does this improve perf? Could you describe a bit what led you to this?

Reply:
I _honestly_ don't know, but using naturally aligned types rather than `#[repr(align)]` significantly impacted performance.

The reason for the specific layout is roughly:

- The first usize should be `*mut` (actually, yeah, it's probably better as `*mut` than `*const`, as `*const` is a lie), because it holds provenance [[I](https://www.ralfj.de/blog/2018/07/24/pointers-and-bytes.html), [II](https://www.ralfj.de/blog/2020/12/14/provenance.html)), and [integers shouldn't hold pointers](https://github.com/rust-lang/rust/issues/95228), [pointer-to-integer transmutes are bad](https://github.com/rust-lang/unsafe-code-guidelines/issues/286), etc
- The rest is to express the "natural" layout of the rest of the type. It could probably be `[u8; PADDING - size_of::<*mut u8>()]` without issue (though I've not tested it).

The reason I tried it at all was because of the ptr-int transmute deal, and it just happened to fix the perf cliff for some unknown reason.

Reply:
Interesting! If we can figure out how/why this improves perf I'm okay with it, but at the moment it feels like any odd compiler change could cause a regression, which I wouldn't want.

I think next steps here are to improve benchmarking and try to answer the question of whether or not the existing perf degradation is because of something like machine code alignment. Reason being, maybe this layout is just helping us get lucky at the moment, and other layouts are just unlucky? I remember seeing a blog post on lobste.rs one time describing how one could go about this, but I'm failing to find it at the moment.

How does this sound to you?

Also kudos for all this work, it's seriously impressive! TIL about `iai` which is a really neat tool! 👏

Reply:
The [link I provided before](https://youtu.be/r-TLSBdHe1A) is to the talk for said tool; unfortunately the tool doesn't work anymore, as it relied heavily on LLVM internals and there wasn't anyone to keep it up to date.

I think the Iai run on the regressed commit shows that it's not just spooky effects like that, as the difference shows up in inst-count as well, which I think is an indication that it's more complicated than that? Either way I've made changes that would perturb code layout and perf has stayed consistent (not all pushed, as they were experimentally to check for resiliency to such in a bodged manner, and if such changes improved things), so I'm reasonably sure it has to do with the layout specifics that rustc/LLVM know about `Repr`, rather than code layout.

Unfortunately, I'm at (or beyond) the end of my knowledge trying to diagnose the perf impact of this, so I'm out of useful tricks to pull out of my hat; I've played all my cards; insert another metaphor here.